zerver/views/home.py

```diff
@@ -102,20 +102,25 @@ def update_last_reminder(user_profile: Optional[UserProfile]) -> None:
 
 
 def home(request: HttpRequest) -> HttpResponse:
-    if not settings.ROOT_DOMAIN_LANDING_PAGE:
-        return home_real(request)
-
+    subdomain = get_subdomain(request)
     # If settings.ROOT_DOMAIN_LANDING_PAGE, sends the user the landing
     # page, not the login form, on the root domain
-
-    subdomain = get_subdomain(request)
-    if subdomain != Realm.SUBDOMAIN_FOR_ROOT_DOMAIN:
-        return home_real(request)
-
-    return hello_view(request)
+    if settings.ROOT_DOMAIN_LANDING_PAGE and subdomain == Realm.SUBDOMAIN_FOR_ROOT_DOMAIN:
+        return hello_view(request)
+
+    # It doesn't matter if the realm is web public or not if
+    # user doesn't wants to login anonymously. We redirect
+    # user to login page in both cases anyway. Plus, we
+    # avoid a database query by checking prefers_web_public_view
+    # first.
+    if (
+        request.POST.get("prefers_web_public_view") == "true"
+        or request.session.get("prefers_web_public_view")
+    ) and get_valid_realm_from_request(request).allow_web_public_streams_access():
+        return web_public_view(home_real)(request)
+    return zulip_login_required(home_real)(request)
```

Comment:
What happens if you make a request to a nonexistent realm with prefers_web_public_view=True on the URL of a realm that doesn't exist?  I'd like to make sure the error isn't a 500.

Reply:
This piece of middleware handles that pretty well:
```
  563
  564 class HostDomainMiddleware(MiddlewareMixin):
  565     def process_request(self, request: HttpRequest) -> Optional[HttpResponse]:
  566     ¦   # Match against ALLOWED_HOSTS, which is rather permissive;
  567     ¦   # failure will raise DisallowedHost, which is a 400.
  568     ¦   request.get_host()
  569
  570     ¦   # This check is important to avoid doing the extra work of
  571     ¦   # `get_realm` (which does a database query that could be
  572     ¦   # problematic for Tornado).  Also the error page below is only
  573     ¦   # appropriate for a page visited in a browser, not the API.
  574     ¦   #
  575     ¦   # API authentication will end up checking for an invalid
  576     ¦   # realm, and throw a JSON-format error if appropriate.
  577     ¦   if request.path.startswith(("/static/", "/api/", "/json/")):
  578     ¦   ¦   return None
  579
  580     ¦   subdomain = get_subdomain(request)
  581     ¦   if subdomain == settings.SOCIAL_AUTH_SUBDOMAIN:
  582     ¦   ¦   # Realms are not supposed to exist on SOCIAL_AUTH_SUBDOMAIN.
  583     ¦   ¦   return None
  584
  585     ¦   request_notes = RequestNotes.get_notes(request)
  586     ¦   try:
  587     ¦   ¦   request_notes.realm = get_realm(subdomain)
  588     ¦   ¦   request_notes.has_fetched_realm = True
  589     ¦   except Realm.DoesNotExist:
  590     ¦   ¦   if subdomain == Realm.SUBDOMAIN_FOR_ROOT_DOMAIN:
  591     ¦   ¦   ¦   # The root domain is used for creating new
  592     ¦   ¦   ¦   # organizations even if it does not host a realm.
  593     ¦   ¦   ¦   return None
  594
  595     ¦   ¦   return render(request, "zerver/invalid_realm.html", status=404)
  596
  597     ¦   return None
```

Reply:
Great, thanks for checking!